File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/PurePursuit/Base/Coordination/Pose.java
package org.firstinspires.ftc.teamcode.PurePursuit.Base.Coordination;

import com.acmerobotics.roadrunner.Pose2d;

public class Pose {

    public volatile double x, y, theta;

    public Pose (Pose pose) {
        this.x = pose.x;
        this.y = pose.y;
        this.theta = pose.theta;
    }

    public Pose (double x, double y, double theta) {
        this.x = x;
        this.y = y;
        this.theta = theta;
    }

    public Pose (Vector vector, double theta) {
        setVec(vector);
        this.theta = theta;
    }

    public void setVec(Vector vector) {
        this.x = vector.x;
        this.y = vector.y;
    }

    public void pose2dTranslation(Pose2d pose2d) {
        this.x = pose2d.position.x;
        this.y = pose2d.position.y;
        this.theta = Math.toDegrees(pose2d.heading.toDouble());
    }

    public Vector getVec() {
        return new Vector(x, y);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/math/Pose.kt
package org.firstinspires.ftc.teamcode.util.math

import com.acmerobotics.roadrunner.geometry.Pose2d
import org.firstinspires.ftc.teamcode.util.Waypoint

data class Pose(
    var p: Point,
    var h: Angle
) {
    val x get() = p.x
    val y get() = p.y
    val cos get() = h.cos
    val sin get() = h.sin
    val hypot get() = p.hypot
    val copy get() = Pose(p, h)

    val pose2d get() = Pose2d(p.x, p.y, h.angle)

    fun distance(p2: Pose) = p.distance(p2.p)
    fun distance(p2: Waypoint) = p.distance(p2.p)

    val toRawString = String.format("%.2f, %.2f, %.2f", x, y, h.angle)
    override fun toString() = String.format("%.2f, %.2f, %.2f", x, y, h.wrap().deg)

    companion object {
        val DEFAULT_ANGLE = Pose(Point.ORIGIN, Angle.EAST)
        val DEFAULT_RAW = Pose(Point.ORIGIN, Angle(0.0, AngleUnit.RAW))
    }
}
